File: csharp/EncodeKata/SessionModificationCmd.cs
using System;

namespace EncodeKata
{
    public class SessionModificationCmd
    {
        private readonly MessageType messageType;
        private readonly int sessionId;
        private readonly int transactionId;
        private bool hasPqvl;
        private int pqvl;
        private SRYP srypId;
        private bool hasXyzTimer;
        private XyzTimer xyzTimer;

        public SessionModificationCmd(int sessionId, int transactionId)
        {
            this.messageType = MessageType.SessionModification;
            this.sessionId = sessionId;
            this.transactionId = transactionId;
            this.xyzTimer = new XyzTimer();
            this.srypId = SRYP.SessionManagement;
        }

        public void SetXyzTimer(XyzTimerUnit unit, int timerValue)
        {
            this.hasXyzTimer = true;
            this.xyzTimer.Set(unit, timerValue);
        }

        public void SetPqvl(int value)
        {
            this.hasPqvl = true;
            this.pqvl = value;
        }   

        public void Encode(ByteBuffer data)
        {
            data.Append((int)this.srypId);
            data.Append(this.sessionId);
            data.Append(this.transactionId);
            data.Append((int)this.messageType);

            if (this.hasXyzTimer)
            {
                this.xyzTimer.Encode(data);
            }
            if (this.hasPqvl)
            {
                this.encodePqvl(data);
            }
        }

        private void encodePqvl(ByteBuffer data)
        {
            data.Append((int)YEW.YewAlwaysOnSession << 4 | (this.pqvl & 0x01));
        }
    }

    internal class XyzTimer
    {
        private XyzTimerUnit timerUnit;
        private int timerValue;

        public void Set(XyzTimerUnit unit, int timerValue)
        {
            this.timerUnit = unit;
            this.timerValue = timerValue;
        }

        public void Encode(ByteBuffer data)
        {
            data.Append((int)YEW.YewXyzTimer);
            int temp;
            temp = (int)this.timerUnit << 5;

            if (this.timerUnit != XyzTimerUnit.TimerDeactivated)
            {
                if (this.timerValue <= 0x1F) // 5 bits max range
                {
                    temp |= (this.timerValue & 0x1F);
                }
                else
                {
                    Console.Write("XYZ timer value out of range. Encoding max value\n");
                    temp |= 0x1F;
                }
            }
            data.Append(temp);
        }
    }
    
    public enum SRYP
    {
        SessionManagement = 0x3,
    }

    public enum YEW
    {
        YewXyzTimer = 0x8,
        YewAlwaysOnSession = 0x9,
    }
    public enum MessageType 
    {
        SessionCreation = 0x0,
        SessionModification = 0x1,
        SessionDelete = 0x2,
    };
    public enum XyzTimerUnit 
    {
        TimerDeactivated,
        MultiplesOfHours,
        MultipliesOfMinutes,
        MultiplesOfSeconds,
    };
}

File: swift/Sources/Encode-TestDesign-Kata/SessionModificationCmd.swift
class SessionModificationCmd {
    private let messageType: MessageType = .sessionModification
    private let sessionId: Int
    private let transactionId: Int
    private var hasPqvl = false
    private var pqvl = 0
    private var srypId: SRYP = .sessionManagement
    private var hasXyzTimer = false
    private var xyzTimer = XyzTimer()

    init(sessionId: Int, transactionId: Int) {
        self.sessionId = sessionId
        self.transactionId = transactionId
    }

    func setXyzTimer(unit: XyzTimerUnit, timerValue: Int) {
        hasXyzTimer = true
        xyzTimer.set(unit: unit, timerValue: timerValue)
    }

    func setPqvl(value: Int) {
        hasPqvl = true
        pqvl = value
    }

    func encode(_ data: ByteBuffer) {
        data.append(srypId.rawValue)
        data.append(sessionId)
        data.append(transactionId)
        data.append(messageType.rawValue)

        if hasXyzTimer {
            xyzTimer.encode(data: data)
        }
        if hasPqvl {
            encodePqvl(data: data)
        }
    }

    private func encodePqvl(data: ByteBuffer) {
        data.append(YEW.yewAlwaysOnSession.rawValue << 4 | (pqvl & 0x01))
    }
}

class XyzTimer {
    private var timerUnit: XyzTimerUnit = .timerDeactivated
    private var timerValue = 0

    func set(unit: XyzTimerUnit, timerValue: Int) {
        timerUnit = unit
        self.timerValue = timerValue
    }

    func encode(data: ByteBuffer) {
    data.append(YEW.yewXyzTimer.rawValue)
        var temp = timerUnit.rawValue << 5

        if timerUnit != .timerDeactivated {
            if timerValue <= 0x1F { // 5 bits max range
                temp |= timerValue & 0x1F
            } else {
                print("XYZ timer value out of range. Encoding max value")
                temp |= 0x1F
            }
        }
        data.append(temp)
    }
}

enum SRYP: Int {
    case sessionManagement = 0x3
}

enum YEW: Int {
    case yewXyzTimer = 0x8
    case yewAlwaysOnSession = 0x9
}

enum MessageType: Int {
    case sessionCreation = 0x0
    case sessionModification = 0x1
    case sessionDelete = 0x2
}

enum XyzTimerUnit: Int {
    case timerDeactivated
    case multiplesOfHours
    case multipliesOfMinutes
    case multiplesOfSeconds
}
